logdna/logdna.py

```diff
@@ -102,21 +103,26 @@ def buffer_log_sync(self, message):
         # Attempt to acquire lock to write to buf
         # otherwise write to secondary as flush occurs
         if self.lock.acquire(blocking=False):
-            msglen = len(message['line'])
-            if self.buf_size + msglen < self.buf_retention_limit:
-                self.buf.append(message)
-                self.buf_size += msglen
-            else:
-                self.internalLogger.debug(
-                    'The buffer size exceeded the limit: %s',
-                    self.buf_retention_limit)
-
-            if self.buf_size >= self.flush_limit and not self.exception_flag:
-                self.close_flusher()
-                self.flush()
-            else:
-                self.start_flusher()
-            self.lock.release()
+            try:
+                msglen = len(message['line'])
+                if self.buf_size + msglen < self.buf_retention_limit:
+                    self.buf.append(message)
+                    self.buf_size += msglen
+                else:
+                    self.internalLogger.debug(
+                        'The buffer size exceeded the limit: %s',
+                        self.buf_retention_limit)
+
+                if self.buf_size >= self.flush_limit and \
+                        not self.exception_flag:
+                    self.close_flusher()
+                    self.flush()
+                else:
+                    self.start_flusher()
+            except Exception as e:
+                self.internalLogger.exception(f'Error in buffer_log_sync: {e}')
+            finally:
+                self.lock.release()
         else:
             self.secondary.append(message)
```

Comment:
It seems like this is still quite vulnerable to a race condition, as the attempt to append to `self.secondary` is not protected by any lock here.

Reply:
@emelski Great catch and you are absolutely right. The idea of using a secondary buffer just strikes me as turtles all the way down for this type of application. It would probably be better to refactor the code to use a simple queue, which is already guaranteed to be thread-safe, rather than the current design of using a lock to synchronize access to a simple list.  Using a queue, we could remove the locking alltogether.

Reply:
To be clear, there is still locking happening behind the scenes, but it's encapsulated within the queue implementation itself, so at the very least it simplifies the code. Not sure if there will be any negative performance regressions with that sort of implementation so I will test it thoroughly.

Reply:
I'm not sure the built-in Queue has the right semantics for the use here -- that is, that Queue is more-or-less designed around the notion of each item in the queue being handled discretely from the others, whereas in our use we want to treat _everything currently in the queue_ as a single block of data to reduce the number of separate requests that are sent to our servers.  Obviously you could implement such semantics using the Queue but it's not quite as simple as saying "let's just use a queue".  In some ways it might be quite clunky to do, in fact, since the Queue would not prohibit threads adding to the queue at just the moment the flushing thread was trying to pull things out to buffer for transmission.  Again, not an unsolvable problem but not an obvious slam dunk replacement either, imo.

Also a minor semantic nit: it's misleading to suggest that using Queue would remove locking altogether.  The locking is still there, it's just pushed into the Queue implementation.  That's easier for our application, to be sure, but does come with caveats -- for example, we'd no longer be able to lock around multiple actions at once, as we do now.  I don't know whether we really _need_ to do that, but it appears that we _do_ do that now, so we'd need to be mindful of the change.

I'm not familiar with the specifics of how the library maintainers have chosen to implement locking in that component, so I can't really speak to whether the performance characteristics and other behaviors of their choices are suitable for our needs here.

Reply:
>I'm not sure the built-in Queue has the right semantics for the use here -- that is, that Queue is more-or-less designed around the notion of each item in the queue being handled discretely from the others, whereas in our use we want to treat everything currently in the queue as a single block of data to reduce the number of separate requests that are sent to our servers. Obviously you could implement such semantics using the Queue but it's not quite as simple as saying "let's just use a queue". In some ways it might be quite clunky to do, in fact, since the Queue would not prohibit threads adding to the queue at just the moment the flushing thread was trying to pull things out to buffer for transmission. Again, not an unsolvable problem but not an obvious slam dunk replacement either, imo.

>Also a minor semantic nit: it's misleading to suggest that using Queue would remove locking altogether. The locking is still there, it's just pushed into the Queue implementation. That's easier for our application, to be sure, but does come with caveats -- for example, we'd no longer be able to lock around multiple actions at once, as we do now. I don't know whether we really need to do that, but it appears that we do do that now, so we'd need to be mindful of the change.

>I'm not familiar with the specifics of how the library maintainers have chosen to implement locking in that component, so I can't really speak to whether the performance characteristics and other behaviors of their choices are suitable for our needs here.

Right. My thought would be that we do something like the following at the time we send data:
```
# Get the size of the queue. Since there is only ever one thread consuming from it this is guaranteed to only either stay the same size or increase during the time this thread is executing
curr_queue_size = self.queue.qsize()

# Consume elements from queue
logs_to_send = [self.queue.get(block=False) for _ in range(curr_queue_size)]

# Continue the flush per existing semantics
self.send_request(logs_to_send)
```

That being said, I see your point and can see how the Queue implementation could possibly yield worse performance due to context switches between threads while handling multiple concurrent calls to `add()` and `get()`. The current implementation has the advantage that the thing that does the draining can consume all of the buffered data in a single locked section.

Reply:
A less drastic change from the current design might be to have a single buffer in our class (doing away with `self.secondary`), with very simple locking just around pushing to it and swapping it with a local buffer in `try_request` or one of its callers.  Something like this:

```
def log(message):
    self.lock.acquire(True)
    self.buf.append(message)
    self.lock.release()

def try_request():
    self.lock.acquire(True)
    if not self.buf:
        self.lock.release()
        return

    localbuf = self.buf
    self.buf = []
    self.lock.release()

    // send the stuff in localbuf
```

Obviously this is not a functional replacement, it's just meant to demonstrate the intent -- the locking protocol is vastly simplified, we still have the ability to manage entire buffers of messages at once without any funny business, and we can continue to accumulate new messages while sending the old.  This is basically how the node agent works in fact, so it's not really a novel design even within our code base.

Reply:
In case it's not clear, I'm not necessarily advocating for one or the other approach.  Both (or even all three?) suggestions on the table have merit, so as the implementer I think the choice is up to you.  But I do think _something_ needs to be done about the race over `self.secondary`, or the problems will just persist, albeit perhaps at a lower rate than originally observed.

Reply:
I like that. I'll take a crack at getting that working.